packages/router/src/grpc/view-protocol-server/balances.ts

```diff
@@ -1,81 +1,162 @@
 import type { Impl } from '.';
 import { servicesCtx } from '../../ctx';
 
-import { addLoHi, Base64Str, uint8ArrayToBase64 } from '@penumbra-zone/types';
+import {
+  addLoHi,
+  Base64Str,
+  getAmountFromValueView,
+  uint8ArrayToBase64,
+} from '@penumbra-zone/types';
 
-import { Value } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1/asset_pb';
-import { AddressIndex } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/keys/v1/keys_pb';
+import {
+  AssetId,
+  ValueView,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1/asset_pb';
+import {
+  AddressIndex,
+  AddressView,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/keys/v1/keys_pb';
+import {
+  AddressByIndexRequest,
+  AssetMetadataByIdRequest,
+  BalancesRequest,
+  BalancesResponse,
+  SpendableNoteRecord,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1/view_pb';
+import { getAssetIdFromRecord } from '@penumbra-zone/types/src/getters/spendable-note-record';
+import { HandlerContext } from '@connectrpc/connect';
+import { assetMetadataById } from './asset-metadata-by-id';
+import { addressByIndex } from './address-by-index';
 import { Amount } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/num/v1/num_pb';
-import { BalancesResponse } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1/view_pb';
-
-type BalancesMap = Record<Base64Str, BalancesResponse>;
-type AccountMap = Record<AddressIndex['account'], BalancesMap>;
 
 // Handles aggregating amounts and filtering by account number/asset id
 export const balances: Impl['balances'] = async function* (req, ctx) {
   const services = ctx.values.get(servicesCtx);
   const { indexedDb } = await services.getWalletServices();
 
-  const accounts: AccountMap = {};
+  const aggregator = new BalancesAggregator(ctx);
 
   for await (const noteRecord of indexedDb.iterateSpendableNotes()) {
     if (noteRecord.heightSpent !== 0n) continue;
+    await aggregator.add(noteRecord);
+  }
+
+  yield* aggregator.filteredResponses(req);
+};
```

Comment:
Instead of one long function, abstracted this logic into a class below. Feels more readable, especially given the expanded functionality.